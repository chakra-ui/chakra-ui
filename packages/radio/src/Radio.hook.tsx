import { useBooleanState, useControllableProp } from "@chakra-ui/hooks"
import { attr, callAllHandlers, Dict, mergeRefs } from "@chakra-ui/utils"
import { visuallyHiddenStyle } from "@chakra-ui/visually-hidden"
import {
  ChangeEvent,
  KeyboardEvent,
  useCallback,
  useRef,
  useState,
} from "react"

/**
 * @todo use the `useTabbable` hook here
 * to manage the isFocusable & isDisabled props
 */
export interface RadioHookProps {
  /**
   * id assigned to input
   */
  id?: string
  /**
   * The name of the input field in a radio
   * (Useful for form submission).
   */
  name?: string
  /**
   * The value to be used in the radio button.
   * This is the value that will be returned on form submission.
   */
  value?: string | number
  /**
   * If `true`, the radio will be checked.
   * You'll need to pass `onChange` to update it's value (since it's now controlled)
   */
  isChecked?: boolean
  /**
   * If `true`, the radio will be initially checked.
   */
  defaultIsChecked?: boolean
  /**
   * If `true`, the radio will be disabled
   */
  isDisabled?: boolean
  /**
   * If `true` and `isDisabled` is true, the radio will remain
   * focusable but not interactive.
   */
  isFocusable?: boolean
  /**
   * If `true`, the radio will be read-only
   */
  isReadOnly?: boolean
  /**
   * If `true`, the radio button will be invalid. This sets `aria-invalid` to `true`.
   */
  isInvalid?: boolean
  /**
   * If `true`, the radio button will be invalid. This sets `aria-invalid` to `true`.
   */
  isRequired?: boolean
  /**
   * Function called when checked state of the `input` changes
   */
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void
}

export function useRadio(props: RadioHookProps = {}) {
  const {
    defaultIsChecked,
    isChecked: isCheckedProp,
    isFocusable,
    isDisabled,
    isReadOnly,
    isRequired,
    onChange,
    isInvalid,
    name,
    value,
    id,
    ...htmlProps
  } = props

  const [isFocused, setFocused] = useBooleanState()
  const [isHovered, setHovering] = useBooleanState()
  const [isActive, setActive] = useBooleanState()

  const ref = useRef<HTMLInputElement>(null)

  const [isCheckedState, setCheckedState] = useState(Boolean(defaultIsChecked))

  const [isControlled, isChecked] = useControllableProp(
    isCheckedProp,
    isCheckedState,
  )

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    if (isReadOnly || isDisabled) {
      event.preventDefault()
      return
    }

    if (!isControlled) {
      setCheckedState(event.target.checked)
    }

    onChange?.(event)
  }

  const trulyDisabled = isDisabled && !isFocusable

  const handleKeyDown = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === " ") {
        setActive.on()
      }
    },
    [setActive],
  )

  const handleKeyUp = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === " ") {
        setActive.off()
      }
    },
    [setActive],
  )

  return {
    state: {
      isInvalid,
      isFocused,
      isChecked,
      isActive,
      isHovered,
      isDisabled,
      isReadOnly,
      isRequired,
    },
    getCheckboxProps: (props: Dict = {}) => ({
      ...props,
      "data-active": attr(isActive),
      "data-hover": attr(isHovered),
      "data-checked": attr(isChecked),
      "data-focus": attr(isFocused),
      "data-readonly": attr(isReadOnly),
      "aria-hidden": true,
      onPointerDown: callAllHandlers(props.onPointerDown, setActive.on),
      onPointerUp: callAllHandlers(props.onPointerUp, setActive.off),
      onPointerEnter: callAllHandlers(props.onPointerEnter, setHovering.on),
      onPointerLeave: callAllHandlers(props.onPointerLeave, setHovering.off),
    }),
    getInputProps: (props: Dict = {}) => ({
      ...props,
      ref: mergeRefs(props.ref, ref),
      type: "radio",
      name,
      value,
      id,
      onChange: callAllHandlers(props.onChange, handleChange),
      onBlur: callAllHandlers(props.onBlur, setFocused.off),
      onFocus: callAllHandlers(props.onFocus, setFocused.on),
      onKeyDown: callAllHandlers(props.onKeyDown, handleKeyDown),
      onKeyUp: callAllHandlers(props.onKeyUp, handleKeyUp),
      "aria-required": attr(isRequired),
      checked: isChecked,
      disabled: trulyDisabled,
      readOnly: isReadOnly,
      "aria-invalid": attr(isInvalid),
      "aria-disabled": attr(isDisabled),
      style: visuallyHiddenStyle,
    }),
    htmlProps,
  }
}

export type RadioHookReturn = ReturnType<typeof useRadio>
