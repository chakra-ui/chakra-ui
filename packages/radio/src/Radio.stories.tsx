import React from "react"
import { useRadio, Radio, useRadioGroup } from "."
import { chakra } from "@chakra-ui/system"
import { Stack } from "@chakra-ui/layout"

export default {
  title: "Radio",
  decorators: [
    (story: Function) => (
      <chakra.div maxWidth="500px" mx="auto" mt="40px">
        {story()}
      </chakra.div>
    ),
  ],
}

export const RadioOnly = () => (
  <Radio marginTop="40px" size="lg" colorScheme="green">
    I'm a radio
  </Radio>
)

function RadioGroup(props: any) {
  const options = ["Option 1", "Option 2", "Option 3"]

  const { getRadioProps, getRootProps } = useRadioGroup(props)

  return (
    <div {...getRootProps()}>
      {options.map(value => (
        <label key={value}>
          <input type="radio" {...getRadioProps({ value })} />
          <span style={{ margin: 10 }}>{value}</span>
        </label>
      ))}
    </div>
  )
}

export const radioGroup = () => (
  <RadioGroup isNative defaultValue={"Option 1"} onChange={console.log} />
)

export const WithTheming = () => {
  const options = ["react", "vue", "svelte"]

  const { getRootProps, getRadioProps } = useRadioGroup({
    name: "test",
    defaultValue: "vue",
    onChange: console.log,
  })

  return (
    <Stack spacing="20px" direction="row" {...getRootProps()}>
      {options.map(value => (
        <Radio key={value} {...getRadioProps({ value })}>
          {value}
        </Radio>
      ))}
    </Stack>
  )
}

function RadioButton(props: any) {
  const { children, ...rest } = props
  const { getInputProps, getCheckboxProps } = useRadio(rest)

  return (
    <chakra.label>
      <input {...getInputProps()} />
      <chakra.div
        {...getCheckboxProps()}
        display="inline-block"
        border="1px solid gray"
        _checked={{ bg: "tomato", color: "white" }}
        _focus={{ outline: "3px dotted red" }}
        px={5}
        py={3}
      >
        {children}
      </chakra.div>
    </chakra.label>
  )
}

export function CustomRadioCard() {
  const options = ["react", "vue", "svelte"]

  const { getRootProps, getRadioProps } = useRadioGroup({
    name: "test",
    defaultValue: "vue",
    onChange: console.log,
  })

  return (
    <Stack direction="row" {...getRootProps()}>
      {options.map(option => (
        <RadioButton {...getRadioProps({ value: option })}>
          {option}
        </RadioButton>
      ))}
    </Stack>
  )
}
