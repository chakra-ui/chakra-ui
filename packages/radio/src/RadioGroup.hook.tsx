import { useControllableProp, useId } from "@chakra-ui/hooks"
import { Dict, isInputEvent, mergeRefs } from "@chakra-ui/utils"
import { ChangeEvent, useCallback, useRef, useState } from "react"

type Value = string | number

type EventOrValue = ChangeEvent<HTMLInputElement> | Value

export interface RadioGroupHookProps {
  /**
   * The value of the radio to be `checked`
   * (in controlled mode)
   */
  value?: Value
  /**
   * The value of the radio to be `checked`
   * initially (in uncontrolled mode)
   */
  defaultValue?: Value
  /**
   * Function called once a radio is checked
   * @param nextValue the value of the checked radio
   */
  onChange?(nextValue: Value): void
  /**
   * The `name` attribute forwarded to each `radio` element
   */
  name?: string
  /**
   * If `true`, input elements will receive
   * `checked` attribute instead of `isChecked`.
   *
   * This assumes, you're using native radio inputs
   */
  isNative?: boolean
  /**
   * The `ref` of the wrapper element of the radio group
   */
  ref?: React.Ref<any>
}

/**
 * useRadioGroup
 *
 * React hook to manage a group of radio inputs
 */
export function useRadioGroup(props: RadioGroupHookProps = {}) {
  const {
    onChange: onChangeProp,
    value: valueProp,
    defaultValue,
    isNative,
  } = props

  const [valueState, setValue] = useState<Value>(defaultValue || "")

  const [isControlled, derivedValue] = useControllableProp(
    valueProp,
    valueState,
  )

  const rootRef = useRef<any>(null)

  const focus = useCallback(() => {
    const rootNode = rootRef.current
    if (!rootNode) return

    let query = `input:not(:disabled):checked`

    const firstEnabledAndCheckedInput = rootNode.querySelector(
      query,
    ) as HTMLElement

    if (firstEnabledAndCheckedInput) {
      firstEnabledAndCheckedInput.focus()
      return
    }

    query = `input:not(:disabled)`

    const firstEnabledInput = rootNode.querySelector(query) as HTMLElement
    firstEnabledInput?.focus()
  }, [])

  /**
   * All radio options must use the same name
   */
  const name = useId(props.name, `radio`)

  const onChange = useCallback(
    (eventOrValue: EventOrValue) => {
      const nextValue = isInputEvent(eventOrValue)
        ? eventOrValue.target.value
        : eventOrValue

      if (!isControlled) {
        setValue(nextValue)
      }

      onChangeProp?.(nextValue)
    },
    [onChangeProp, isControlled],
  )

  return {
    actions: {
      focus,
      setValue,
    },
    state: { value: derivedValue },
    getRadioProps: (props: Dict = {}) => ({
      ...props,
      name,
      onChange,
      value: props.value,
      ...(isNative
        ? { checked: props.value === derivedValue }
        : { isChecked: props.value === derivedValue }),
    }),
    getRootProps: (props: Dict = {}) => ({
      ...props,
      ref: mergeRefs(props.ref, rootRef),
      role: "radiogroup",
    }),
  }
}
