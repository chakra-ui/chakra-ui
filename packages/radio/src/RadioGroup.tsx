import * as React from "react"
import {
  __DEV__,
  getValidChildren,
  mapResponsive,
  omit,
} from "@chakra-ui/utils"
import { useRadioGroup, RadioGroupHookProps } from "./RadioGroup.hook"
import { chakra, useTheme, css, SystemProps, PropsOf } from "@chakra-ui/system"

export type RadioGroupProps = RadioGroupHookProps &
  Omit<PropsOf<typeof chakra.div>, "onChange" | "value"> & {
    /**
     * The name of the radio group
     */
    name?: string
    /**
     * The space between the children radios
     */
    spacing?: SystemProps["margin"]
    /**
     * The direction to stack the children radios
     */
    direction?: SystemProps["flexDirection"]
  }

/**
 * RadioGroup
 *
 * Used for multiple radios which are bound in one group,
 * and it indicates which option is selected.
 *
 * @see Docs https://chakra-ui.com/radio
 *
 */

export const RadioGroup = React.forwardRef(
  (props: RadioGroupProps, ref: React.Ref<any>) => {
    const {
      name,
      colorScheme,
      size,
      spacing = 2,
      direction = "row",
      children,
      ...rest
    } = props
    const theme = useTheme()
    const validChildren = getValidChildren(children)
    const { getRootProps, getRadioProps } = useRadioGroup({ ...props, ref })

    const childSpacing = mapResponsive(spacing, value => {
      const { margin } = css({ margin: value })(theme)
      return `calc(${margin} / 2)`
    })

    const containerSpacing = mapResponsive(spacing, value => {
      const { margin } = css({ margin: value })(theme)
      return `calc(${margin} / 2 * -1)`
    })

    const clones = validChildren.map((child, index) => {
      return (
        <chakra.div key={index} margin={childSpacing}>
          {React.cloneElement(child, getRadioProps(child.props))}
        </chakra.div>
      )
    })

    return (
      <chakra.div
        display="flex"
        flexWrap="wrap"
        role="radiogroup"
        {...getRootProps()}
        flexDirection={direction}
        margin={containerSpacing}
        {...omit(rest, ["onChange"])}
      >
        {clones}
      </chakra.div>
    )
  },
)

if (__DEV__) {
  RadioGroup.displayName = "RadioGroup"
}
