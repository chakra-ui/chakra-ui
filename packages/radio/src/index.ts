export * from "./Radio"
export * from "./Radio.hook"
export * from "./RadioGroup.hook"
export * from "./RadioGroup"
