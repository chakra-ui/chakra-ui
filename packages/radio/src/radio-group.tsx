import { ThemingProps, chakra, PropsOf, forwardRef } from "@chakra-ui/system"
import { createContext, __DEV__, cx } from "@chakra-ui/utils"
import * as React from "react"
import {
  useRadioGroup,
  UseRadioGroupProps,
  UseRadioGroupReturn,
} from "./use-radio-group"

export interface RadioGroupContext
  extends Pick<UseRadioGroupReturn, "onChange" | "value" | "name">,
    Omit<ThemingProps, "orientation"> {}

const [RadioGroupProvider, useRadioGroupContext] = createContext<
  RadioGroupContext
>({
  name: "RadioGroupContext",
  strict: false,
})

export { useRadioGroupContext }

export type RadioGroupProps = UseRadioGroupProps &
  Omit<
    PropsOf<typeof chakra.div>,
    "onChange" | "value" | "defaultValue" | "children"
  > &
  Omit<ThemingProps, "orientation"> & { children: React.ReactNode }

/**
 * Used for multiple radios which are bound in one group,
 * and it indicates which option is selected.
 *
 * @see Docs https://chakra-ui.com/components/radio
 */
export const RadioGroup = forwardRef(function RadioGroup(
  props: RadioGroupProps,
  ref: React.Ref<any>,
) {
  const {
    colorScheme,
    size,
    variant,
    children,
    className,
    ...otherProps
  } = props

  const { value, onChange, getRootProps, name, htmlProps } = useRadioGroup(
    otherProps,
  )

  const group = React.useMemo(
    () => ({
      name,
      size,
      onChange,
      colorScheme,
      value,
      variant,
    }),
    [size, name, onChange, colorScheme, value, variant],
  )

  const groupProps = getRootProps(htmlProps, ref)
  const _className = cx("chakra-radio-group", className)

  return (
    <RadioGroupProvider value={group}>
      <chakra.div {...groupProps} className={_className}>
        {children}
      </chakra.div>
    </RadioGroupProvider>
  )
})

if (__DEV__) {
  RadioGroup.displayName = "RadioGroup"
}
