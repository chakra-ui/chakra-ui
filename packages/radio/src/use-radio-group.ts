import { useControllableProp, useId } from "@chakra-ui/hooks"
import { StringOrNumber, isInputEvent } from "@chakra-ui/utils"
import * as React from "react"
import { mergeRefs, PropGetter } from "@chakra-ui/react-utils"

type EventOrValue = React.ChangeEvent<HTMLInputElement> | StringOrNumber

export interface UseRadioGroupProps {
  /**
   * The value of the radio to be `checked`
   * (in controlled mode)
   */
  value?: StringOrNumber
  /**
   * The value of the radio to be `checked`
   * initially (in uncontrolled mode)
   */
  defaultValue?: StringOrNumber
  /**
   * Function called once a radio is checked
   * @param nextValue the value of the checked radio
   */
  onChange?(nextValue: string): void
  /**
   * If `true`, all wrapped radio inputs will be disabled
   */
  isDisabled?: boolean

  /**
   * If `true` and `isDisabled` is true, all wrapped radio inputs will remain
   * focusable but not interactive.
   */
  isFocusable?: boolean
  /**
   * The `name` attribute forwarded to each `radio` element
   */
  name?: string
  /**
   * If `true`, input elements will receive
   * `checked` attribute instead of `isChecked`.
   *
   * This assumes, you're using native radio inputs
   */
  isNative?: boolean
}

type RadioPropGetter = PropGetter<
  HTMLInputElement,
  {
    onChange?: (e: EventOrValue) => void
    value?: StringOrNumber
    /**
     * checked is defined if isNative=true
     */
    checked?: boolean
    /**
     * isChecked is defined if isNative=false
     */
    isChecked?: boolean
  } & Omit<
    React.InputHTMLAttributes<HTMLInputElement>,
    "onChange" | "size" | "value"
  >
>

/**
 * React hook to manage a group of radio inputs
 */
export function useRadioGroup(props: UseRadioGroupProps = {}) {
  const {
    onChange: onChangeProp,
    value: valueProp,
    defaultValue,
    name: nameProp,
    isDisabled,
    isFocusable,
    isNative,
    ...htmlProps
  } = props

  const [valueState, setValue] = React.useState<StringOrNumber>(
    defaultValue || "",
  )
  const [isControlled, value] = useControllableProp(valueProp, valueState)

  const ref = React.useRef<any>(null)

  const focus = React.useCallback(() => {
    const rootNode = ref.current
    if (!rootNode) return

    let query = `input:not(:disabled):checked`

    const firstEnabledAndCheckedInput = rootNode.querySelector(
      query,
    ) as HTMLElement

    if (firstEnabledAndCheckedInput) {
      firstEnabledAndCheckedInput.focus()
      return
    }

    query = `input:not(:disabled)`

    const firstEnabledInput = rootNode.querySelector(query) as HTMLElement
    firstEnabledInput?.focus()
  }, [])

  /**
   * All radio options must use the same name
   */
  const fallbackName = useId(undefined, `radio`)
  const name = nameProp || fallbackName

  const onChange = React.useCallback(
    (eventOrValue: EventOrValue) => {
      const nextValue = isInputEvent(eventOrValue)
        ? eventOrValue.target.value
        : eventOrValue

      if (!isControlled) {
        setValue(nextValue)
      }

      onChangeProp?.(String(nextValue))
    },
    [onChangeProp, isControlled],
  )

  const getRootProps: PropGetter = React.useCallback(
    (props = {}, forwardedRef = null) => ({
      ...props,
      ref: mergeRefs(forwardedRef, ref),
      role: "radiogroup",
    }),
    [],
  )

  const getRadioProps: RadioPropGetter = React.useCallback(
    (props = {}, ref = null) => {
      const checkedKey = isNative ? "checked" : "isChecked"
      return {
        ...props,
        ref,
        name,
        [checkedKey]: value != null ? props.value === value : undefined,
        onChange,
        "data-radiogroup": true,
      }
    },
    [isNative, name, onChange, value],
  )

  return {
    getRootProps,
    getRadioProps,
    name,
    ref,
    focus,
    setValue,
    value,
    onChange,
    isDisabled,
    isFocusable,
    htmlProps,
  }
}

export type UseRadioGroupReturn = ReturnType<typeof useRadioGroup>
