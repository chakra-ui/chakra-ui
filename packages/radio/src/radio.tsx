import {
  chakra,
  layoutPropNames,
  PropsOf,
  SystemProps,
  ThemingProps,
  useMultiStyleConfig,
  omitThemingProps,
} from "@chakra-ui/system"
import { cx, split, __DEV__ } from "@chakra-ui/utils"
import * as React from "react"
import { useRadioGroupContext } from "./radio-group"
import { useRadio, UseRadioProps } from "./use-radio"

type Omitted = "onChange" | "defaultChecked" | "checked"
type BaseControlProps = Omit<PropsOf<typeof chakra.div>, Omitted>

export type RadioProps = UseRadioProps &
  ThemingProps &
  BaseControlProps & {
    /**
     * The spacing between the checkbox and it's label text
     * @default 0.5rem
     */
    spacing?: SystemProps["marginLeft"]
    /**
     * If `true`, the radio will occupy the full width of it's parent container
     */
    isFullWidth?: boolean
  }

/**
 * Radio component is used in forms when a user needs to select a single value from
 * several options.
 * @see Docs https://chakra-ui.com/components/radio
 */
export const Radio = React.forwardRef(function Radio(
  props: RadioProps,
  ref: React.Ref<any>,
) {
  const group = useRadioGroupContext()
  const styles = useMultiStyleConfig("Radio", { ...group, ...props })

  const {
    spacing = "0.5rem",
    children,
    isFullWidth,
    ...radioProps
  } = omitThemingProps(props)

  let isChecked = props.isChecked
  if (group?.value && props.value) {
    isChecked = group.value === props.value
  }

  let onChange = props.onChange
  if (group?.onChange && props.value) {
    onChange = group.onChange
  }

  const name = group?.name || props?.name

  const {
    getInputProps,
    getCheckboxProps,
    getLabelProps,
    htmlProps: rest,
  } = useRadio({ ...radioProps, isChecked, onChange, name })

  const [layoutProps, otherProps] = split(rest, layoutPropNames as any)

  const checkboxProps = getCheckboxProps(otherProps)
  const inputProps = getInputProps({}, ref)
  const labelProps = getLabelProps()

  const rootStyles = {
    width: isFullWidth ? "full" : undefined,
    display: "inline-flex",
    alignItems: "center",
    verticalAlign: "top",
  }

  const checkboxStyles = {
    display: "inline-flex",
    alignItems: "center",
    justifyContent: "center",
    flexShrink: 0,
    ...styles.control,
  }

  const labelStyles = {
    userSelect: "none",
    ml: spacing,
    ...styles.label,
  }

  return (
    <chakra.label className="chakra-radio" {...layoutProps} __css={rootStyles}>
      <input className="chakra-radio__input" {...inputProps} />
      <chakra.div
        className="chakra-radio__control"
        {...checkboxProps}
        __css={checkboxStyles}
      />
      {children && (
        <chakra.div
          className="chakra-radio__label"
          {...labelProps}
          __css={labelStyles}
          children={children}
        />
      )}
    </chakra.label>
  )
})

if (__DEV__) {
  Radio.displayName = "Radio"
}
