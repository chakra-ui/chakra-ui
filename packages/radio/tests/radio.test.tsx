import * as React from "react"
import { render, fireEvent } from "@chakra-ui/test-utils"
import { useRadio, UseRadioProps } from "../src"

test("has proper aria and data attributes", async () => {
  const Component = (props: UseRadioProps = {}) => {
    const { getCheckboxProps, getInputProps } = useRadio(props)

    return (
      <label>
        <input data-testid="input" {...getInputProps()} />
        <div data-testid="checkbox" {...getCheckboxProps()} />
      </label>
    )
  }
  const utils = render(<Component name="name" value="" id="id" />)

  let input = utils.getByTestId("input")
  let checkbox = utils.getByTestId("checkbox")

  expect(input).toHaveAttribute("name", "name")
  expect(input).toHaveAttribute("id", "id")
  expect(input).toHaveAttribute("value", "")
  expect(input).not.toBeDisabled()
  expect(input).not.toHaveAttribute("aria-required")
  expect(input).not.toHaveAttribute("aria-invalid")
  expect(input).not.toHaveAttribute("aria-disabled")
  expect(checkbox).toHaveAttribute("aria-hidden", "true")
  expect(checkbox).not.toHaveAttribute("data-active")
  expect(checkbox).not.toHaveAttribute("data-hover")
  expect(checkbox).not.toHaveAttribute("data-checked")
  expect(checkbox).not.toHaveAttribute("data-focus")
  expect(checkbox).not.toHaveAttribute("data-readonly")

  // render with various flags enabled
  utils.rerender(<Component isDisabled isInvalid isReadOnly isRequired />)

  input = utils.getByTestId("input")
  checkbox = utils.getByTestId("checkbox")

  expect(input).toHaveAttribute("aria-required")
  expect(input).toHaveAttribute("aria-invalid")
  expect(input).toHaveAttribute("aria-disabled")
  expect(input).toBeDisabled()
  expect(checkbox).toHaveAttribute("data-readonly")

  // input is not truly disabled if focusable
  utils.rerender(<Component isDisabled isFocusable />)

  input = utils.getByTestId("input")

  expect(input).not.toBeDisabled()
})

test("handles events and callbacks correctly", () => {
  const hookProps = { onChange: jest.fn() }
  const checkboxProps = {
    onMouseDown: jest.fn(),
    onMouseUp: jest.fn(),
  }
  const inputProps = {
    onChange: jest.fn(),
    onBlur: jest.fn(),
    onFocus: jest.fn(),
    onKeyDown: jest.fn(),
    onKeyUp: jest.fn(),
  }
  const Component = () => {
    const { getCheckboxProps, getInputProps } = useRadio(hookProps)

    return (
      <label>
        <input data-testid="input" {...getInputProps(inputProps)} />
        <div data-testid="checkbox" {...getCheckboxProps(checkboxProps)} />
      </label>
    )
  }
  const utils = render(<Component />)
  const input = utils.getByTestId("input")
  const checkbox = utils.getByTestId("checkbox")

  // mouse up and down
  fireEvent.mouseDown(checkbox)
  expect(checkbox).toHaveAttribute("data-active")
  expect(checkboxProps.onMouseDown).toHaveBeenCalled()

  fireEvent.mouseUp(checkbox)
  expect(checkbox).not.toHaveAttribute("data-active")
  expect(checkboxProps.onMouseUp).toHaveBeenCalled()

  // on change
  fireEvent.click(input)
  expect(input).toBeChecked()
  expect(checkbox).toHaveAttribute("data-checked")
  expect(hookProps.onChange).toHaveBeenCalled()
  expect(inputProps.onChange).toHaveBeenCalled()

  // blur and focus
  fireEvent.focus(input)
  expect(checkbox).toHaveAttribute("data-focus")
  expect(inputProps.onFocus).toHaveBeenCalled()

  fireEvent.blur(input)
  expect(checkbox).not.toHaveAttribute("data-focus")
  expect(inputProps.onFocus).toHaveBeenCalled()

  // key down and key up
  fireEvent.keyDown(input, { key: " ", keyCode: 32 })
  expect(checkbox).toHaveAttribute("data-active")
  expect(inputProps.onKeyDown).toHaveBeenCalled()

  fireEvent.keyUp(input, { key: " ", keyCode: 32 })
  expect(checkbox).not.toHaveAttribute("data-active")
  expect(inputProps.onKeyUp).toHaveBeenCalled()
})
