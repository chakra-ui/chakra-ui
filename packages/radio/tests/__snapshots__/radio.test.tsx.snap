// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders correctly 1`] = `
<DocumentFragment>
  .emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: top;
}

.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 1.25rem;
  -webkit-transition: box-shadow 250ms,background-color 250ms;
  transition: box-shadow 250ms,background-color 250ms;
  border: 2px solid;
  border-radius: 9999px;
  border-color: inherit;
  color: #FFFFFF;
  height: 1.25rem;
}

.emotion-0[aria-checked=true],
.emotion-0[data-checked] {
  background: #3182CE;
  border-color: #3182CE;
  color: #FFFFFF;
}

.emotion-0[aria-checked=true]:hover,
.emotion-0[data-checked]:hover,
.emotion-0[aria-checked=true][data-hover],
.emotion-0[data-checked][data-hover] {
  background: #2A69AC;
  border-color: #2A69AC;
}

.emotion-0[aria-checked=true][disabled],
.emotion-0[data-checked][disabled],
.emotion-0[aria-checked=true][disabled]:focus,
.emotion-0[data-checked][disabled]:focus,
.emotion-0[aria-checked=true][disabled]:hover,
.emotion-0[data-checked][disabled]:hover,
.emotion-0[aria-checked=true][aria-disabled=true],
.emotion-0[data-checked][aria-disabled=true],
.emotion-0[aria-checked=true][aria-disabled=true]:focus,
.emotion-0[data-checked][aria-disabled=true]:focus,
.emotion-0[aria-checked=true][aria-disabled=true]:hover,
.emotion-0[data-checked][aria-disabled=true]:hover,
.emotion-0[aria-checked=true][data-disabled],
.emotion-0[data-checked][data-disabled],
.emotion-0[aria-checked=true][data-disabled]:focus,
.emotion-0[data-checked][data-disabled]:focus,
.emotion-0[aria-checked=true][data-disabled]:hover,
.emotion-0[data-checked][data-disabled]:hover {
  border-color: #E2E8F0;
  background: #E2E8F0;
  color: #718096;
}

.emotion-0[aria-checked=true]::before,
.emotion-0[data-checked]::before {
  content: "";
  display: inline-block;
  position: relative;
  width: 50%;
  height: 50%;
  border-radius: 50%;
  background: currentColor;
}

.emotion-0:indeterminate,
.emotion-0[aria-checked=mixed],
.emotion-0[data-indeterminate] {
  background: #3182CE;
  border-color: #3182CE;
  color: #FFFFFF;
}

.emotion-0[disabled],
.emotion-0[disabled]:focus,
.emotion-0[disabled]:hover,
.emotion-0[aria-disabled=true],
.emotion-0[aria-disabled=true]:focus,
.emotion-0[aria-disabled=true]:hover,
.emotion-0[data-disabled],
.emotion-0[data-disabled]:focus,
.emotion-0[data-disabled]:hover {
  background: #EDF2F7;
  border-color: #EDF2F7;
}

.emotion-0:focus,
.emotion-0[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-0[aria-invalid=true],
.emotion-0[data-invalid] {
  border-color: #E53E3E;
}

<label
    class="chakra-radio emotion-1"
  >
    <input
      class="chakra-radio__input"
      style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
      type="radio"
      value=""
    />
    <div
      aria-hidden="true"
      class="chakra-radio__control emotion-0"
    />
  </label>
</DocumentFragment>
`;
